internal/parse/parser.go

```diff
@@ -243,58 +236,64 @@ func (p *Parser) skipName() bool {
 //  - bool == false, err == nil
 //		The construct was not the one we are looking for
 
-// parseIdentifier parses either a name made up only of nameBytes or an
-// asterisk.
-func (p *Parser) parseIdentifier() (string, bool) {
-	if p.skipByte('*') {
-		return "*", true
-	}
+type asteriskFlag int
+
+const (
+	allowAsterisk asteriskFlag = iota
+	disallowAsterisk
+)
 
-	idStart := p.pos
+// parseIdentifier parses a name made up of only nameBytes, or alternatively a
+// single asterisk if `allowAsterisk` is passed. On success it returns the
+// parsed string and true. Otherwise, it returns the empty string and false.
+func (p *Parser) parseIdentifier(asteriskF asteriskFlag) (string, bool) {
+	if asteriskF == allowAsterisk {
+		if p.skipByte('*') {
+			return "*", true
+		}
+	}
+	mark := p.pos
 	if p.skipName() {
-		return p.input[idStart:p.pos], true
+		return p.input[mark:p.pos], true
 	}
 	return "", false
 }
 
-// parseFullName parses a column name or Go field name, optionally dot-prefixed by
-// its table or type name respectively.
-func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+// parseGoObject parses a source or target go object of the form Prefix.Name
+// where the type is the Prefix and the field is the Name (this applies to maps
+// and structs). On success it returns the parsed FullName, true and nil. If a
+// Go object is found, but not formatted correctly, false and an error are
+// returned. Otherwise the error is nil.
+func (p *Parser) parseGoObject() (FullName, bool, error) {
 	cp := p.save()
 	var fn FullName
-	if id, ok := p.parseIdentifier(); ok {
-		fn.Prefix = id
+	if id, ok := p.parseIdentifier(disallowAsterisk); ok {
 		if p.skipByte('.') {
-			if id, ok := p.parseIdentifier(); ok {
-				fn.Name = id
-				return fn, true
+			if idField, ok := p.parseIdentifier(allowAsterisk); ok {
+				return FullName{id, idField}, true, nil
 			}
-		} else {
-			// A column name specified without a table prefix is a name not a
-			// prefix.
-			if idc == columnId {
-				fn.Name = fn.Prefix
-				fn.Prefix = ""
-			}
-			return fn, true
+			return fn, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
+		return fn, false, fmt.Errorf("go object near char %d not qualified", p.pos)
 	}
 	cp.restore()
-	return fn, false
+	return fn, false, nil
```

Comment:
Could this method return `nil, false, nil` in failure cases instead of `fn, false, nil` because with this change, `fn` seems to be no longer used. We could delete the declaration for `fn` at the top as well :)

Reply:
Thats a good point. Go doesnt actually allow us to use nil there because its not a zero value for structs, but I have just changed it to return `FullName{}` instead so I can remove the var declaration.